Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+    }
+
+    auto gbmBuffer = makeUnique<MemoryMappedGPUBuffer>(size);
+    if (!gbmBuffer->allocate(device, bufferFormat.value())) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->map()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to map GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->createEGLImageFromDMABuf()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to export GBM buffer as EGLImage");
+        return nullptr;
+    }
+
+    return gbmBuffer;
+}
+
+bool MemoryMappedGPUBuffer::allocate(struct gbm_device* device, const GLDisplay::DMABufFormat& bufferFormat)
+{
+    m_modifier = DRM_FORMAT_MOD_INVALID;
+    if (bufferFormat.modifiers.size() > 0)
+        m_bo = gbm_bo_create_with_modifiers2(device, m_size.width(), m_size.height(), bufferFormat.fourcc, bufferFormat.modifiers.data(), bufferFormat.modifiers.size(), GBM_BO_USE_RENDERING);
+
+    if (m_bo) {
+        m_modifier = gbm_bo_get_modifier(m_bo);
+
+        if (dumpBufferAllocation) {
+            if (m_bo)
+                LOG(Images, " -> buffer allocated with modifiers (gbm_bo reports modifier 0x%02lx)", m_modifier);
+        }
+    } else {
+        m_bo = gbm_bo_create(device, m_size.width(), m_size.height(), bufferFormat.fourcc, GBM_BO_USE_LINEAR);
+        m_modifier = DRM_FORMAT_MOD_INVALID;
+
+        if (dumpBufferAllocation) {
+            if (m_bo)
+                LOG(Images, " -> buffer allocated without modifiers (gbm_bo reports modifier 0x%2lx, m_modifier 0x%2lx)", gbm_bo_get_modifier(m_bo), m_modifier);
+        }
+    }
+
+    if (m_bo) {
+        if (!exportGBMBufferObjectAsDMABuf())
+            return false;
+
+        if (dumpBufferAllocation)
+            LOG(Images, " -> exporting gbm_bo buffer handle %u as dma-buf, got fd %i! succeeded!", gbm_bo_get_handle(m_bo).u32, m_dmaBufFD);
+
+        return true;
+    }
+
+    return false;
+}
+
+bool MemoryMappedGPUBuffer::map()
+{
+    ASSERT(!m_mappedData);
+    ASSERT(!m_mappedLength);
+    m_stride = gbm_bo_get_stride_for_plane(m_bo, 0);
+    m_mappedLength = m_stride * m_size.height();
+    m_mappedData = mmap(nullptr, m_mappedLength, PROT_READ | PROT_WRITE, MAP_SHARED, m_dmaBufFD, 0);
+    return m_mappedData != MAP_FAILED;
+}
+
+void MemoryMappedGPUBuffer::unmap()
+{
+    ASSERT(m_mappedData);
+    ASSERT(m_mappedLength > 0);
+    munmap(m_mappedData, m_mappedLength);
+}
+
+bool MemoryMappedGPUBuffer::createEGLImageFromDMABuf()
+{
+    auto planeCount = gbm_bo_get_plane_count(m_bo);
+    if (planeCount != 1) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] No support for multi-plane buffer objects. Linear, single plane buffers are necessary.");
+        return false;
+    }
+
+    auto dmaBufFD = gbm_bo_get_fd_for_plane(m_bo, 0);
```

Comment:
I am similarly tempted to suggest a RAII object for the file descriptor. I see that `UnixFileDescriptor` exists, could we use it here?

Reply:
Done.